### PR TITLE
Add /get_ppl endpoint

### PR DESCRIPTION
**Pull request by irexyc**
## Usage
```
curl -X POST 'http://127.0.0.1:23333/get_ppl' \
    -H 'Content-Type: application/json' \
    -d '{
      "input": [9707, 11, 1246, 525, 498, 3351, 30]
    }'


curl -X POST 'http://127.0.0.1:23333/get_ppl' \
    -H 'Content-Type: application/json' \
    -d '{"input": "Hello, how are you today? Hello"}'
```

## Accuracy 
| dataset | version | metric | mode | old pytorch | new pytorch | Δ pytorch | old turbomind | new turbomind | Δ turbomind |
|---|---|---|---|---:|---:|---:|---:|---:|---:|
| race-high | 9387ad | accuracy | ppl | 62.75 | 62.75 | 0.00 | 62.61 | 62.72 | +0.11 |
| GPQA_diamond | 4b5a83 | accuracy | ppl | 31.31 | 30.81 | -0.50 | 32.32 | 31.31 | -1.01 |
| winogrande | 252f01 | accuracy | ll | 57.06 | 57.06 | 0.00 | 57.38 | 57.38 | 0.00 |
| mmlu-other | - | accuracy | ppl | 49.74 | 49.74 | 0.00 | 49.83 | 49.47 | -0.36 |

### CODE REVIEW

File: lmdeploy/pipeline.py
Comment (lvhan028) [inline lmdeploy/pipeline.py]: `_get_ppl` and  `_batch_iterator` are dead code now. Please remove them

File: lmdeploy/pytorch/engine/model_agent/agent.py [hunk 1]
```diff
@@ -653,6 +654,14 @@ async def _step_postprocess_with_output(self,
             extra_inputs=extra_inputs,
         )
 
+        # compute summed, unnormalized prompt cross-entropy
+        ce_loss = None
+        if return_ce_loss and logits is not None:
```
Comment (grimoire) [inline lmdeploy/pytorch/engine/model_agent/agent.py]: `last_logits` is the inputs of async_sampling_logits, and it is also a slice of logits. In sampling last_logits might be updated inplace.

File: lmdeploy/serve/openai/protocol.py [hunk 2]
```diff
@@ -448,6 +447,29 @@ class PoolingResponse(BaseModel):
     usage: UsageInfo
 
 
+class GetPPLRequest(BaseModel):
+    """Get perplexity request.
+
+    The input may be raw text or token ids. Text is tokenized with
+    ``tokenizer.encode`` (no chat template applied).
+    """
+    model: str | None = None
```
Comment (lvhan028) [inline lmdeploy/serve/openai/protocol.py]: Can `model` be None?

File: lmdeploy/serve/openai/protocol.py [hunk 3]
```diff
@@ -448,6 +447,29 @@ class PoolingResponse(BaseModel):
     usage: UsageInfo
 
 
+class GetPPLRequest(BaseModel):
```
Comment (lvhan028) [inline lmdeploy/serve/openai/protocol.py]: Let's rename it "PPLRequest"

File: lmdeploy/serve/openai/protocol.py [hunk 4]
```diff
@@ -448,6 +447,29 @@ class PoolingResponse(BaseModel):
...
+    """Get perplexity request.
+
+    The input may be raw text or token ids. Text is tokenized with
+    ``tokenizer.encode`` (no chat template applied).
+    """
+    model: str | None = None
+    input: str | list[int]
+
+
+class GetPPLResponse(BaseModel):
```
Comment (lvhan028) [inline lmdeploy/serve/openai/protocol.py]: `PPLResponse`

File: lmdeploy/serve/openai/protocol.py [hunk 5]
```diff
@@ -448,6 +447,29 @@ class PoolingResponse(BaseModel):
...
+    """Get perplexity response.
+
+    ``ppl`` is the perplexity (mean cross-entropy loss) of the input.
+    """
+    id: str = Field(default_factory=lambda: f'ppl-{shortuuid.random()}')
+    object: str = 'ppl'
+    created: int = Field(default_factory=lambda: int(time.time()))
+    model: str = None
+    ppl: float
+    usage: UsageInfo
```
Comment (lvhan028) [inline lmdeploy/serve/openai/protocol.py]: Since `usage` play an important role in 'ppl' endpint, may remove it.

File: lmdeploy/serve/openai/api_server.py [hunk 6]
```diff
@@ -1198,6 +1182,43 @@ async def pooling(request: PoolingRequest, raw_request: Request = None):
...
+
+    - **model** (str): model name. Available from /v1/models.
+    - **input** (str | list[int]): the input to score, either raw text or token
+      ids. Text is tokenized with ``tokenizer.encode`` (no chat template is
+      applied).
+    """
+    async_engine = VariableInterface.async_engine
+
+    request_input = request.input
+    model_name = request.model or async_engine.model_name
```
Comment (lvhan028) [inline lmdeploy/serve/openai/api_server.py]: Let's validate request.model

File: lmdeploy/serve/core/async_engine.py [hunk 7]
```diff
@@ -883,3 +853,54 @@ async def _proc(session, i):
...
+            raise ValueError('input must have at least 2 tokens to compute ppl.')
+
+        ce_loss = None
+        session = self.session_mgr.get()
+        try:
+            async with session.request_handle() as handle:
+                # TurboMind needs one decode token to drive the request to
+                # FINISH. The in-engine CE reduction still scores prompt tokens
+                # only.
+                max_new_tokens = 1 if self.backend == 'turbomind' else 0
```
Comment (lvhan028) [inline lmdeploy/serve/core/async_engine.py]: Do we have to differentiate the engines now? Can't we use the same max_new_tokens(1 is preferred) for both engines?

Comment (lvhan028) [reply lmdeploy/serve/core/async_engine.py]: Also, in `async_get_logits` there is also such case. May take it into consideration

File: lmdeploy/pipeline.py [hunk 8]
```diff
@@ -283,40 +281,12 @@ def get_ppl(self, input_ids: list[int] | list[list[int]]) -> list[float]:
...
-                result.extend(res)
-                for session in sessions:
-                    self.session_mgr.remove(session)
-        output = list(range(len(result)))
-        for index, sorted_index in enumerate(indices):
-            output[sorted_index] = result[index]
-        return output
+        engine = self.async_engine
+        async def _gather():
+            return await asyncio.gather(*[engine.async_get_ppl(ids) for ids in input_ids])
```
Comment (lvhan028) [inline lmdeploy/pipeline.py]: Batch `get_ppl` concurrency is unbounded.
Consider reusing the limiter as `Pipeline.infer` does.
```python
async def _gather():
    sem = self._get_limiter()

    async def _one(ids):
        await sem.acquire()
        try:
            return await engine.async_get_ppl(ids)
        finally:
            sem.release()

    return await asyncio.gather(*[_one(ids) for ids in input_ids])
```

**Review by grimoire — APPROVED** [review]

**Review by lzhangzz — APPROVED** [review]
